File: internal/names/names.go
package names

import (
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/lolopinto/ent/internal/codegen/codegenapi"
)

// TODO even through strcase has been updated. time to own this
// so it doesn't change on us again

// TODO need a better name for this

func ToClassType(strs ...string) string {
	var sb strings.Builder
	for _, s := range strs {
		if !utf8.ValidString(s) {
			sb.WriteString(s)
			continue
		}
		for _, v := range splitCamelCase(s) {
			if isSeparatorToken(v.entry) {
				continue
			}
			if v.class == upper || v.class == digit {
				sb.WriteString(v.entry)
			} else {
				sb.WriteString(titleWord(v.entry))
			}
		}
	}
	return sb.String()
}

func ToTsFieldName(strs ...string) string {
	var sb strings.Builder
	hasDoneLower := false
	for idx, s := range strs {
		if idx == 0 && len(s) > 0 && s[0] == '_' {
			s = s[1:]
			sb.WriteString("_")
		}
		if s == "" {
			continue
		}
		if !utf8.ValidString(s) {
			sb.WriteString(s)
			hasDoneLower = true
			continue
		}

		var prev caseType = not_returned
		tokens := splitCamelCase(s)
		for i, v := range tokens {
			if isSeparatorToken(v.entry) {
				prev = v.class
				continue
			}
			if !hasDoneLower {
				sb.WriteString(lowerWord(v.entry))
				hasDoneLower = true
			} else if v.class == upper || v.class == digit {
				sb.WriteString(titleWord(v.entry))
			} else if v.class == other && prev == upper && len(v.entry) > 2 {
				sb.WriteString(lowerWord(v.entry))
			} else if v.class == lower && prev == upper && i == len(tokens)-1 && v.entry == "s" {
				sb.WriteString(v.entry)
			} else {
				sb.WriteString(titleWord(v.entry))
			}
			prev = v.class
		}
	}
	return sb.String()
}

func ToGraphQLName(cfg codegenapi.Config, s ...string) string {
	// special case id
	if len(s) == 1 && strings.ToLower(s[0]) == "id" {
		return "id"
	}
	if cfg.DefaultGraphQLFieldFormat() == codegenapi.LowerCamelCase {
		return ToTsFieldName(s...)
	}
	return ToDBColumn(s...)
}

func ToGraphQLNameIgnoreSettings(s ...string) string {
	if len(s) == 1 && strings.ToLower(s[0]) == "id" {
		return "id"
	}
	return ToTsFieldName(s...)

}

// ToDBColumn converts a string to a database column name
func ToDBColumn(strs ...string) string {
	var sb strings.Builder
	for idx, s := range strs {
		if idx == 0 && len(s) > 0 && s[0] == '_' {
			s = s[1:]
			sb.WriteString("_")
		}
		if s == "" {
			continue
		}
		sb.WriteString(toSnakeWord(s))

		if idx != len(strs)-1 {
			sb.WriteString("_")
		}
	}
	return sb.String()
}

// same implementation as ToDBColumn
func ToFilePathName(s string) string {
	// file path in file system
	// snake case
	return ToDBColumn(s)
}

func ToGraphQLEnumName(s string) string {
	// norm for graphql enum names is all caps
	return strings.ToUpper(ToDBColumn(s))
}

type caseType uint

const (
	lower caseType = iota
	upper
	digit
	other
	not_returned
)

type splitResult struct {
	entry string
	class caseType
}

type splitTempResult struct {
	runes []rune
	v     caseType
}

// lifted from https://github.com/fatih/camelcase/blob/master/camelcase.go
// and https://github.com/fatih/camelcase/pull/4/files
func splitCamelCase(s string) []splitResult {
	if !utf8.ValidString(s) {
		return []splitResult{
			{
				entry: s,
				class: other,
			},
		}
	}
	// var runes []struct{[]rune, caseType} = []struct{[]rune, caseType}{}
	var temp []splitTempResult
	// lastClass := 0
	// class := 0
	var lastClass caseType = not_returned
	var class caseType = not_returned

	// split into fields based on class of unicode character
	for _, r := range s {
		switch true {
		case unicode.IsLower(r):
			class = lower
		case unicode.IsUpper(r):
			class = upper
		case unicode.IsDigit(r):
			class = digit
		default:
			class = other
		}
		if class == lastClass {
			temp[len(temp)-1].runes = append(temp[len(temp)-1].runes, r)
		} else {
			temp = append(temp, splitTempResult{
				runes: []rune{r},
				v:     class,
			})
		}
		lastClass = class
	}

	// this is for handling the userIDs -> "user", "ID", "s" case
	isPlural := func(curr, next []rune) bool {
		return len(curr) > 1 && len(next) == 1 && next[0] == 's'
	}

	// handle upper case -> lower case, number --> lower case sequences, e.g.
	// "PDFL", "oader" -> "PDF", "Loader"
	// "192", "nd" -> "192nd", ""
	for i := 0; i < len(temp)-1; i++ {
		if unicode.IsUpper(temp[i].runes[0]) && unicode.IsLower(temp[i+1].runes[0]) && !isPlural(temp[i].runes, temp[i+1].runes) {
			temp[i+1].runes = append([]rune{temp[i].runes[len(temp[i].runes)-1]}, temp[i+1].runes...)
			temp[i].runes = temp[i].runes[:len(temp[i].runes)-1]

			temp[i+1].v = other
		} else if unicode.IsDigit(temp[i].runes[0]) && unicode.IsLower(temp[i+1].runes[0]) {
			temp[i].runes = append(temp[i].runes, temp[i+1].runes...)
			temp[i+1].runes = nil
			i++

			temp[i].v = other
		}
	}

	results := []splitResult{}

	// construct []string from results
	for _, s := range temp {
		if len(s.runes) > 0 {
			results = append(results, splitResult{
				class: s.v,
				entry: string(s.runes),
			})
		}
	}
	return results
}

File: ts/src/names/names.ts
// TODO tests and update functionality to match golang
function splitWords(input: string): string[] {
  if (!input) {
    return [];
  }

  const parts = input.split(/[^A-Za-z0-9]+/).filter(Boolean);
  const words: string[] = [];

  for (const part of parts) {
    const split = _splitCamelCase(part);
    for (const piece of split) {
      if (!piece.s) {
        continue;
      }
      if (piece.type === "digit" && words.length > 0) {
        words[words.length - 1] += piece.s;
      } else {
        words.push(piece.s);
      }
    }
  }

  return words;
}

function upperFirst(input: string): string {
  if (!input) {
    return "";
  }
  return input[0].toUpperCase() + input.slice(1);
}

function camelCase(input: string): string {
  const words = splitWords(input);
  if (words.length === 0) {
    return "";
  }
  const [first, ...rest] = words;
  return (
    first.toLowerCase() +
    rest.map((word) => upperFirst(word.toLowerCase())).join("")
  );
}

function pascalCase(input: string): string {
  return splitWords(input)
    .map((word) => upperFirst(word.toLowerCase()))
    .join("");
}

function snakeCase(input: string): string {
  return splitWords(input)
    .map((word) => word.toLowerCase())
    .join("_");
}

export function toDBColumnOrTable(...strs: string[]): string {
  let name = "";
  for (let i = 0; i < strs.length; i++) {
    const s = strs[i];

    // this is to handle userIDs -> user_ids
    const split = _splitCamelCase(s);
    if (split.length > 2) {
      const last = split[split.length - 1];
      const nextLast = split[split.length - 2];
      if (last.s === "s" && nextLast.type === "upper") {
        // get the first n-2 words
        name += snakeCase(
          split
            .map((v) => v.s)
            .slice(0, split.length - 2)
            .join(""),
        );
        name += "_";

        // combine the last two
        name += snakeCase(nextLast.s);
        name += last.s;
        continue;
      }
    }
    name += snakeCase(s);
    if (i !== strs.length - 1) {
      name += "_";
    }
  }
  return name;
}

export function toFilePath(s: string): string {
  return snakeCase(s);
}

export function toFieldName(...strs: string[]): string {
  let name = "";
  let hasDoneLower = false;

  for (const s of strs) {
    // same logic in toDBColumnOrTable
    const split = _splitCamelCase(s);
    if (split.length > 2) {
      const last = split[split.length - 1];
      const nextLast = split[split.length - 2];
      if (last.s === "s" && nextLast.type === "upper") {
        // get the first n-2 words
        name += camelCase(
          split
            .map((v) => v.s)
            .slice(0, split.length - 2)
            .join(""),
        );

        // combine the last two
        name += pascalCase(nextLast.s);
        name += last.s;
        hasDoneLower = true;
        continue;
      }
    }
    if (!hasDoneLower) {
      name += camelCase(s);
      hasDoneLower = true;
    } else {
      name += pascalCase(s);
    }
  }
  return name;
}

export function toClassName(str: string): string {
  return pascalCase(str);
}

interface splitResult {
  s: string;
  type: "lower" | "upper" | "digit" | "other";
}

function isUpper(s: string): boolean {
  for (const c of s) {
    if (!(c >= "A" && c <= "Z")) {
      return false;
    }
  }
  return true;
}

function isLower(s: string): boolean {
  for (const c of s) {
    if (!(c >= "a" && c <= "z")) {
      return false;
    }
  }
  return true;
}

function isDigit(s: string): boolean {
  return /^\d+$/.test(s);
}

// ported from golang in internal/names/names.go
export function _splitCamelCase(s: string): splitResult[] {
  const results: splitResult[] = [];

  let lastType: splitResult["type"] | undefined;
  let type: splitResult["type"] | undefined;

  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    if (isLower(c)) {
      type = "lower";
    } else if (isUpper(c)) {
      type = "upper";
    } else if (isDigit(c)) {
      type = "digit";
    } else {
      type = "other";
    }

    if (lastType == type) {
      results[results.length - 1].s += c;
    } else {
      results.push({ s: c, type });
    }
    lastType = type;
  }

  // this is for handling the userIDs -> "user", "ID", "s" case
  const isPlural = function (curr: string, next: string): boolean {
    return curr.length > 1 && next === "s";
  };

  for (let i = 0; i < results.length - 1; i++) {
    if (
      isUpper(results[i].s[0]) &&
      isLower(results[i + 1].s[0]) &&
      !isPlural(results[i].s, results[i + 1].s)
    ) {
      // get last item from results[i] and add to front of results[i+1]
      results[i + 1].s =
        results[i].s[results[i].s.length - 1] + results[i + 1].s;
      // now has a prefix and is not all of one case
      results[i + 1].type = "other";
      // remove last character from results[i]
      results[i].s = results[i].s.slice(0, results[i].s.length - 1);
    } else if (isDigit(results[i].s[0]) && isLower(results[i + 1].s[0])) {
      results[i].s = results[i].s + results[i + 1].s;
      results[i].type = "other";

      // remove
      results[i + 1].s = "";
      i++;
    }
  }
  for (let i = 0; i < results.length; i++) {
    if (results[i].s === "") {
      results.splice(i, 1);
      i--;
    }
  }
  return results;
}
